hugr-passes/src/static_circuit.rs

```diff
@@ -0,0 +1,398 @@
+use std::{
+    collections::{BTreeSet, HashMap, HashSet},
+    hash::{Hash, Hasher},
+    ops::Deref,
+};
+
+use ascent::Lattice;
+use hugr_core::{
+    extension::{prelude::QB_T, ExtensionRegistry, ExtensionSet},
+    ops::{ExtensionOp, NamedOp as _},
+    HugrView, IncomingPort, Node, OutgoingPort, PortIndex, Wire,
+};
+use itertools::{zip_eq, Itertools};
+use petgraph::algo::toposort;
+
+use crate::{
+    dataflow::{
+        AbstractValue, ConstLoader, DFContext, Machine, NoDefaultConversionToSum, PartialValue, Sum,
+    },
+    validation::ValidationLevel,
+};
+
+#[derive(Eq, PartialEq, Debug, Clone, PartialOrd, Ord)]
+pub struct Gate {
+    // The nodes that execute this gate. They must form a "co-cycle" i.e. every
+    // path from input to output contains at most one of these nodes
+    nodes: BTreeSet<Node>,
+    gate: String,
+    gating_nodes: BTreeSet<Node>,
+    hash: u64,
+}
+
+impl Gate {
+    pub fn show(&self) -> String {
+        format!(
+            "{} [{}] ({})",
+            &self.gate,
+            self.gating_nodes.iter().join(","),
+            self.nodes.iter().join(",")
+        )
+    }
+
+    fn calc_hash(&mut self) {
+        let mut hasher = std::hash::DefaultHasher::default();
+        self.nodes.hash(&mut hasher);
+        self.gate.hash(&mut hasher);
+        self.gating_nodes.hash(&mut hasher);
+        self.hash = hasher.finish()
+    }
+
+    pub fn new(
+        node: Node,
+        gate: impl Into<String>,
+        gating_nodes: impl IntoIterator<Item = Node>,
+    ) -> Self {
+        let mut g = Self {
+            nodes: BTreeSet::from_iter([node]),
+            gate: gate.into(),
+            gating_nodes: BTreeSet::from_iter(gating_nodes),
+            hash: 0,
+        };
+        g.calc_hash();
+        g
+    }
+    pub fn join(mut self, other: Self) -> Option<Self> {
+        if self.gate != other.gate || self.gating_nodes != other.gating_nodes {
```

Comment:
So we require `gating_nodes` to be exactly equal?

Doesn't this rule out cases like a conditional where each Case contains "the same" circuit, as the Node IDs will be different? You'd need some kind of isomorphism....

Reply:
I think gating_nodes gets the node-id of the gate added, not the node id of the qalloc (/function input)?

Reply:
Ah I see - yes much clearer now you've refactored `interpret_leaf_op` a bit